### ISSUE TITLE
SAPI support?

### ISSUE THREAD

**Issue by MingliangT** [krakjoe/parallel#35]
Hi @krakjoe 

Thank you for writing parallel! I just can't wait to try it.

> pthreads v3 is restricted to operating in CLI only... So I'm promoting the advice to hard and fast fact: you can't use pthreads safely and sensibly anywhere but CLI.

and is parallel support SAPI (php-fpm)?

Thank you.

**Comment by krakjoe** [comment]
I think I'll probably refer back to this answer in the future, so it will not be short ... sorry about that ...

The reasons that pthreads was restricted to cli are many, among them:

  - no SAPI provides an API to ensure the safety of threads running within other threads, at least, they are not prepared for multiple threads to be writing to the same outputs, in both FPM and Apache this caused random faults when threads started in a web request write the standard output.
  - the threads that pthreads creates are extremely heavy, at least as heavy as the process that started the thread, it copies all code by default.
  - the way pthreads copies code is very wasteful, you might even say flawed ... duplicating large chunks of memory, having to set vm handlers to cleanup the mess we made when preparing the thread

With all of that in mind, consider the following ... If someone said to you they have a PHP script or front end controller that calls shell_exec to start bash, which in turns starts a php process (with no opcaching enabled or possible) which executes a php script, they might start 4 or 8 or 16 such processes in response to every request to their frontend controller ... you would tell them they are doing it wrong, if you could, _you would make up rules to try and stop them_ ...

Problem 1: Output Faults
----------------------------------

So the first thing that needs fixing is the output problem, quite simple really, parallel makes output atomic - all calls to echo/print etc are exclusive, problem solved (no more faults, tested).

Problem 2: Heavy Threads
------------------------------------

parallel and pthreads work very differently when it comes to moving around code between threads, as mentioned, pthreads copies everything and does a pretty poor job of it, it duplicates all classes and functions for every thread started. 

pthreads was invented before anyone thought of preloading, the PHP of the day did not support sharing code between threads, before 7.2, it simply was not safe. pthreads had to make deep copies of all code it shared among threads, for every thread it started, the same is true for opcode caches (apc/opcache). While it is faster to copy code into a requests memory than it is to compile the code for that request, nevertheless with a large application with many thousands of classes it is not without cost in terms of speed, and memory consumption. It also makes a thread not remarkably different to a process in terms of resources used.

parallel does not copy any code when it starts a thread, it only copies that which will be executed ... follows are some notes on how it does that ...

Problem 3: Sharing (and Copying)  Code
------------------------------------------------------

parallel doesn't care about classes, they always come from opcache or are provided by an autoloader, so it never has to copy any classes at all.

parallel has to have a way to buffer and share closures, and do so in a way that is as close to preloading in terms of efficiency as possible.

The two basic problems here are sharing code - that is to say, making code safe to share among threads just as opcache does. It may not be obvious, but the code that is produced by zend compiler is not thread safe, it is opcache that makes code safe to share normally. The second problem is copying code - each thread started has no access to code compiled in the thread that started it by default, parallel, or opcache, has to make it available as quickly as possible.

In the worst case scenario, when opcache is not loaded or a function is not cached for some reason: parallel has it's own cache which produces cached code indistinguishable from opcache. The cache included in parallel should be considered a fallback - if you care about performance, use opcache+parallel. Even so, it's quite efficient, not quite as efficient as opcache itself, but it's not bad at all.

In the best case scenario, when opcache is loaded, _parallel does not need to copy any code at all._, it's already safe to share.

It may sound complex for parallel to have it's own opcode cache, however, the opcode cache is less than 500 lines of code, and it actually simplifies things for parallel - at layers above the cache, code is much simpler because it only ever has to deal with cached code.

So, when parallel comes to buffer a closure, it is certain that the code is allocated in persistent memory, but the closure itself is a function and an object, which may have static properties that differ from the definition of the closure (which is the code in the cache). So the sending side buffers blocks of memory sizeof(zend_function) (that's 224 bytes on x64) in a process wide space (once for each closure being shared), the receiving side takes the 224 byte block and copies it to request memory (once for each closure being executed).

If you gave me a week, I couldn't come up with comparable numbers for pthreads, I don't know by what amount this improves the efficiency of parallel compared to pthreads, but let's call it a bagillion percent.

The threads that parallel executes couldn't really be any lighter while still initialising an interpreter.

_Side note:_
parallel has one eye firmly on the future, in fact I am developing it against PHP 8 with JIT enabled: parallel can and will execute JIT'd code, pthreads can't, and won't ever be able to execute JIT'd code.

TL;DR
---------

You can use parallel in any SAPI that is thread safe: I have tested in apache and fpm a little, everything seems fine. That is not to claim that all of the core programming of fpm, or any SAPI is actually prepared for this kind of execution, but given the guarantees that ZTS provides, and given the highly compatible nature of parallel and opcache, you should be able to move forward with some confidence.

There is still the question of whether creating threads at the frontend of your website makes sense, creating threads in addition to that which the web server must create may interfere with the ability of that server to scale as intended.

I think that's all there is to say about that ...

**Comment by remicollet** [comment]
Notice: FPM is a "Process Manager", so using "fork" model, without thread usage, so usually FPM is build without ZTS support (in most distribution)

**Comment by MingliangT** [comment]
Thank you so much! The way you explain is very clear :)

We running application using nginx & php-fpm. In order to reduce response time, i want to turn some serial RPCs(e.g, http, thrift, mysql) into parallel. 

The sample code is as follows:

```
$resp1 = doSomeHTTP($arg1);
$resp2 = doSomeMySQL($arg2);
$resp3 = doSomeThrift($arg3);
$ret   = doSomething($resp1, $resp2, $resp3);
```

and i want to call functions(doSomeHTTP, doSomeMySQL, doSomeThrift) in parallel.

I have read [ExecutorService](https://gist.github.com/krakjoe/3c4efb20e14db01bb0b9bc88dd14ff7c) and I will try to implement this use parallel extension, which is very useful.

I'm also reading [amphp/parallel-functions](https://github.com/amphp/parallel-functions) ([use parallel internal](https://github.com/amphp/parallel/blob/master/lib/Context/Parallel.php#L177)), maybe it can meets my needs too.

Thanks.
